zap_otlp_encoder/zap_otlp_encoder.go

```diff
@@ -0,0 +1,262 @@
+package zap_otlp_encoder
+
+// adapted from https://github.com/uber-go/zap/blob/master/zapcore/json_encoder.go
+// and https://github.com/uber-go/zap/blob/master/zapcore/console_encoder.go
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"io"
+	"sync"
+	"time"
+
+	"go.opentelemetry.io/otel/trace"
+	v1 "go.opentelemetry.io/proto/otlp/common/v1"
+	lpb "go.opentelemetry.io/proto/otlp/logs/v1"
+	"go.uber.org/zap/buffer"
+	"go.uber.org/zap/zapcore"
+	"google.golang.org/protobuf/proto"
+)
+
+const _hex = "0123456789abcdef"
+
+var bufferPool = buffer.NewPool()
+
+var _otlpPool = sync.Pool{New: func() interface{} {
+	return &otlpEncoder{}
+}}
+
+func getOTLPEncoder() *otlpEncoder {
+	return _otlpPool.Get().(*otlpEncoder)
+}
+
+func putOTLPEncoder(enc *otlpEncoder) {
+	enc.EncoderConfig = nil
+	enc.buf = nil
+	enc.spaced = false
+	enc.openNamespaces = 0
+	enc.reflectBuf = nil
+	enc.reflectEnc = nil
+	enc.log = nil
+	_otlpPool.Put(enc)
+}
+
+type otlpEncoder struct {
+	*zapcore.EncoderConfig
+	buf            *buffer.Buffer
+	spaced         bool
+	openNamespaces int
+
+	log *lpb.LogRecord
+
+	// for encoding generic values by reflection
+	reflectBuf *buffer.Buffer
```

Comment:
**buf**  :-  this is the buffer that zap expects from an encoder.
**log**  :- since we want to store in in otel log format instead of manually adding all the data to buffer, I am adding it to the **log** and then finally converting the log to buffer which zap expects

Reply:
Not using the reflectedBuf for now so have commented it out.

I am fully not clear on why it is required in the json encoder present here https://github.com/uber-go/zap/blob/382e2511e51cda8afde24f9e6e741f934308edfa/zapcore/json_encoder.go#L61